internal/git/git.go

```diff
@@ -0,0 +1,234 @@
+	pass := compare.Pass{}
+	r, err := git.PlainOpenWithOptions(path, &git.PlainOpenOptions{
+		DetectDotGit:          true,
+		EnableDotGitCommonDir: true,
+	})
+	if err != nil {
+		return pass, err
+	}
+	fs := NewGitFS(path, r, false)
+	fsFrom := NewGitFS(path, r, true)
+
+	changed, err := findChangedThrift(path)
+	if err != nil {
+		return pass, err
+	}
+	var errs error
+	for _, c := range changed {
+		var toModule *compile.Module
+		if c.change == merkletrie.Modify {
+			toModule, err = compile.Compile(c.file, compile.Filesystem(fs))
+			if err != nil {
+				return pass, err
+			}
+		} else if c.change == merkletrie.Delete {
+			// something got deleted, so we are creating an empty module here.
+			toModule = &compile.Module{
+				Name: c.file,
+			}
+		}
+
+		fromModule, err := compile.Compile(c.file, compile.Filesystem(fsFrom))
+		if err != nil {
+			return pass, err
+		}
+		pass.Modules(fromModule, toModule)
+	}
+
+	// p will have lints as a field which we can sort in cli.
+
+	return pass, errs
+}
+
+// FS holds reference to components needed for git FS.
+type FS struct {
+	repo   *git.Repository
+	gitDir string
+	from   bool // Whether we are looking for previous version.
+}
+
+func (fs FS) findCommit() (*object.Commit, error) {
+	// Default is look at HEAD.
+	refHead, err := fs.repo.Head()
+	if err != nil {
+		return nil, err
+	}
+	hash := refHead.Hash()
+
+	// Return first commit.
+	if !fs.from {
+		return fs.repo.CommitObject(hash)
+	}
+
+	// Otherwise, we need to look at provided sha.
+	commitIter, err := fs.repo.Log(&git.LogOptions{From: hash})
+	if err != nil {
+		return nil, err
+	}
+	// Skip one.
+	_, err = commitIter.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	return commitIter.Next()
+}
+
+func (fs FS) Read(path string) ([]byte, error) {
+	commit, err := fs.findCommit()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// filename is going to be the full path. We don't want that.
+	filename, err := filepath.Rel(fs.gitDir, path)
+	if err != nil {
+		return nil, err
+	}
+	// It's possible that file was deleted and it will not exist.
+	f, err := commit.File(filename)
+	if err != nil {
+		return nil, err
+	}
+	s, err := f.Contents()
+	if err != nil {
+		return nil, err
+	}
+	body := []byte(s)
+
+	return body, nil
+}
+
+// Abs returns absolute path to a file.
+func (fs FS) Abs(p string) (string, error) {
+	// Sometimes p can be a full path already on includes, and sometimes it can be relative.
+	if filepath.IsAbs(p) {
+		return p, nil
+	}
+
+	return filepath.Join(fs.gitDir, p), nil
+}
+
+type change struct {
+	file   string
+	change merkletrie.Action
+}
+
+// findChangedThrift reads a git repo and finds any Thrift files that got changed
+// between HEAD and previous commit.
+func findChangedThrift(gitDir string) ([]*change, error) {
+	// Init git repo.
+	r, err := git.PlainOpenWithOptions(gitDir, &git.PlainOpenOptions{
+		DetectDotGit:          true,
+		EnableDotGitCommonDir: true,
+	})
```

Comment:
we already have a repo open by the time we call this. can't we pass that repository object here?